File: JavaScript/Medium/trees/validate_BST.js
/**
 * @param {TreeNode} root
 * @return {boolean}
 * 
 */

var TreeNode = (val, left, right) => {
    this.val = (val === undefined ? 0 : val)
    this.right = (right === undefined ? null : right)
    this.left = (left === undefined ? null : left)
}

var isValidBST = function (root) {
    var helper = function (node, lower, upper){
        if (!node) return true;

        if ((lower !== null && node.val <= lower) || (upper !==null && node.val >= upper)) return false;

        if (!helper(node.right, node.val, upper)) return false;
        if (!helper(node.left, lower, node.val)) return false;

        return true
    }
  
    return helper(root, null, null)
}

File: Python/Medium/trees/validate_bst.py



class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val 
        self.left = left 
        self.right = right


class Solution(object):
    def isValidBST(self, root):
        def helper(node, lower, upper):
            if not node: return True

            if lower is not None and node.val <= lower or upper is not None and node.val >=upper: return False

            if not helper(node.right, node.val, upper): return False
            if not helper(node.left, lower, node.val): return False

            return True
        
        return helper(root, None, None)